### ISSUE TITLE
Access credential security breach.

### ISSUE THREAD

**Issue by valschmidt** [mickael-kerjean/filestash#456]
# Description of the bug

`filestash` is applying the credentials of one backend accessed in browser tab 1, to that of a second backend accessed in browser tab 2 when both access the same endpoint but with access IDs and secret keys for users that have been granted different permissions. Let me explain 

# Step by step instructions to reproduce the bug

Create an S3 bucket with two user accounts. Give user1 an S3 policy that grants user1 full read/write permissions. Give user2 an S3 policy that grants user2 read-only permissions in the bucket. 

In `filestash` create two S3 backends to access the bucket. Give them different labels to distinguish them, and apply user1's account credentials to the first and user2's account credentials to the second.

In a new browser tab, access filestash and access the S3 backend for user2 and test that you cannot create a file/folder in the bucket. You should have read-only permissions.
In another browser tab, access filestash and and access the S3 backend for user1 and test that you can create a file/folder in the bucket. You should have read/write permissions.
Return to the previous browser tab where you previously access the S3 backend for user2, and test that you still cannot create a file/folder in the bucket. YOU'LL FIND THAT YOU CAN! The process of accessing the S3 bucket for user1 in the other browser tab has transferred the credentials from user1 to the browser tab for user2's access. The effect is to give user2 complete access.  

# Can you replicate that error from the demo?

No, but only because I can't configure multiple backends for multiple users in this way.

# Observed behavior

I've got one S3 bucket with two user accounts to access it (This is on Wasabi, but I don't think it matters.). One account ("field") has full read/write access. The other, ("public") has read only access. Each has its own access ID/secret key pair.

In the `backends` section of filestash I've configured two S3 backends to access this S3 bucket, one with the field account credentials and the second with the public account credentials. The point here is to serve up two ways to access the same data, one with full read/write access and the second with read-only access. In this preliminary test I've included the secret keys for both accounts in their respective configurations, but in the future I would remove the secret key from the field account force users to enter it when they desire write access.

To test this, I have two browser tabs open to file stash. 

In browser tab A, I select the S3 backend using the "field" account that provides full read/write access. I can test that indeed I do have read-write access.
In browser tab B, I select the S3 backend using the "public" account that provides read-only access. I can test that indeed I do have read-only access.
Then I select browser tab A again, and if I test write access to the bucket it will now fail. The process of selecting the read-only backend in browser tab B, is now allowing me to access the bucket in browser tab A under the restricted, public account. 
Then in browser tab A, I return to the page for selecting the backend (by pressing the power on/off icon in the upper right), and I again select the S3 backend using the "field" account that provides full read/write access. I test and indeed I do now have write access.
Finally, I select browser tab B, where I had previously selected the S3 backend using the "public" account for read-only access, and I find that I now have FULL READ/WRITE ACCESS there.  

# Expected behavior

Permissions within one S3 backend should not affect permissions to another S3 backend accessed through another browser tab.

**Comment by mickael-kerjean** [comment]
Let me know if I'm wrong but it seems you are making the assumption that tabs are isolated between each other. If that's the case, that assumption isn't correct. When you connect on facebook in tab1, you don't need to reconnect to it in tab2, your cookie / session data is shared across all your tabs.

Filestash works in the same way, when you login from tab1, tab2 is now the same as tab1 and if you login from tab2, tab1 session is now the same as tab2. Ultimatly cookies and session data isn't isolated to a particular tab but shared across all your tabs. That's how cookie works and I don't see how it would constitute in a security breach in any way

**Comment by valschmidt** [comment]
Hm. You may be right.

In my experience on other sites, I've sometimes logged in under different credentials in different tabs and in that case the credentials were *not* shared across the tabs. But maybe this was done with another technology? Something different than cookies? I'm not sure. Isolation between tabs was the behavior I was expecting, and when it didn't happen in my testing I was surprised that I suddenly had write access where I previously had only read access.  

I'm sorry for the fire alarm. If this is the behavior that you expect/want, then feel free to close this issue. 

Cheers.

**Comment by mickael-kerjean** [comment]
Yes you'd need to store your session data another way that's local to the tab (localstorage, indexdb, ...). The issue with such an approach is if somebody was to execute code within Filestash (which would constitue a security breach indeed), that person would be able to leak out your session cookie and use that to connect as you. 

In the current implementation, in the event a security breach were to happen like this, the hacker wouldn't be able to leak out your session cookie as it's not accessible from javascript.

Nothing is perfect, just pro and cons
